sdk/metric/meter.go

```diff
@@ -440,52 +450,58 @@ func (m *meter) RegisterCallback(f metric.Callback, insts ...metric.Observable)
 		// Don't allocate a observer if not needed.
 		return noopRegister{}, nil
 	}
-
-	reg := newObserver()
+	unregs := make([]func(), len(m.pipes))
 	var err error
-	for _, inst := range insts {
-		switch o := inst.(type) {
-		case int64Observable:
-			if e := o.registerable(m); e != nil {
-				if !errors.Is(e, errEmptyAgg) {
-					err = errors.Join(err, e)
+	for ix, pipe := range m.pipes {
+		reg := newObserver(pipe)
+		for _, inst := range insts {
+			switch o := inst.(type) {
+			case int64Observable:
+				if e := o.registerable(m); e != nil {
+					if !errors.Is(e, errEmptyAgg) {
+						err = errors.Join(err, e)
+					}
+					continue
 				}
-				continue
-			}
-			reg.registerInt64(o.observableID)
-		case float64Observable:
-			if e := o.registerable(m); e != nil {
-				if !errors.Is(e, errEmptyAgg) {
-					err = errors.Join(err, e)
+				reg.registerInt64(o.observableID)
+			case float64Observable:
+				if e := o.registerable(m); e != nil {
+					if !errors.Is(e, errEmptyAgg) {
+						err = errors.Join(err, e)
+					}
+					continue
 				}
-				continue
+				reg.registerFloat64(o.observableID)
+			default:
+				// Instrument external to the SDK.
+				return nil, fmt.Errorf("invalid observable: from different implementation")
 			}
-			reg.registerFloat64(o.observableID)
-		default:
-			// Instrument external to the SDK.
-			return nil, fmt.Errorf("invalid observable: from different implementation")
 		}
-	}
 
-	if reg.len() == 0 {
-		// All insts use drop aggregation or are invalid.
-		return noopRegister{}, err
+		if reg.len() == 0 {
+			// All insts use drop aggregation or are invalid.
+			return noopRegister{}, err
+		}
+
+		// Some or all instruments were valid.
+		cBack := func(ctx context.Context) error { return f(ctx, reg) }
+		unregs[ix] = pipe.addMultiCallback(cBack)
 	}
 
-	// Some or all instruments were valid.
-	cback := func(ctx context.Context) error { return f(ctx, reg) }
-	return m.pipes.registerMultiCallback(cback), err
+	return m.pipes.registerMultiCallbacks(unregs), err
 }
 
 type observer struct {
 	embedded.Observer
 
+	pipe    *pipeline
 	float64 map[observableID[float64]]struct{}
```

Comment:
If you take my advice above to do validation once for all pipelines, you should be able to remove these maps from the observer.  IIRC, they are only used for validation, so you could just create these maps at the beginning of RegisterCallback.

Reply:
@dashpole For this change, I noticed that these maps are also getting used in ObserveInt64 and ObserveFloat64 function for making sure that the observable is registered or not. So I haven't removed this.

Reply:
makes sense, thanks